on demand caching bug fix

scotmoor opened the pull request.

Commit:
instance caching bug fix

scotmoor commented:
@JargoonPard @milismsft PTAL

milismsft commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/servergroup/cache/AzureServerGroupCachingAgent.groovy:
```diff
@@ -246,7 +126,7 @@ class AzureServerGroupCachingAgent extends AzureCachingAgent {
 
     try {
-      updatedServerGroup = metricsSupport.readData {
+      newServerGroup = metricsSupport.readData {
```
it looks like the code always assumes a new server group. what if the OnDemand is on a delete (see eviction)? If the resource is not there we might get a CloudException or the API call if we are lucky might just return null.

milismsft commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/servergroup/cache/AzureServerGroupCachingAgent.groovy:
```diff
@@ -255,69 +135,114 @@ class AzureServerGroupCachingAgent extends AzureCachingAgent {
+
+    if (cacheResult.cacheResults.values().flatten().isEmpty()) {
+      // evict it
+      providerCache.evictDeletedItems(AZURE_ON_DEMAND.ns, [serverGroupKey])
```
What if a cache aget is in the middle of the execution when the "delete" OnDemand event happens? It will basically restore back the "old" server group, right?

milismsft commented:
The OnDemand  I expect it to work as following:
- figure out first if it's an update or eviction (aka if resource does not exist)
- capture the current time when we post the call to Azure
- check all current keys in the cache provider for the regular entries and if the timestamp is newer register  the entry as an OnDemand update or delete in the cache

On the normal agent run:
- read all the resources from Azure and capture the current time
- check the OnDemand entries, both delete and update
- if the current regular entry is older than the ondemand update entry, use the OnDemand entry to update the regular cache entries
- if the current regular entry is older than the onDemand delete entry, skip it from the update of the regular cache
- for any OnDemand entries, both delete and update, that are older than the current time stamp, evict them from the OnDemand cache.

Couple things we need to keep in mind while implementing this:
- multiple users can pound on the Deck/orca (see pipelines) at the same time; one could create while
other could delete/update. 
- there could be multiple instances of the clouddriver service, though they all expect to use the same redis database.

ajordens committed:
rx-ified the ReservationReportCachingAgent

- Each account will be loaded in parallel

scotmoor commented:
@JargoonPard @milismsft PTAL

ajordens committed:
Merge pull request #487 from ajordens/optimized-reservation-report-caching-agent

rx-ified the ReservationReportCachingAgent

JargoonPard commented:
LGTM

milismsft commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy:
```diff
@@ -148,8 +148,12 @@ public class AzureComputeClient extends AzureBaseClient {
+      }
+      else {
+        throw e
+      }
```
Are we 100% sure we get a 404 CloudException if the resource is not there? I've observed different behavior when trying to delete a non-existing resource in some cases; basically there was no exception.

I'm asking because I'd like to take this code and add it into how we handle the delete of an Azure resource.

Commit:
Fixes issue with order of GCE httpHealthCheck and instance health check calls.

milismsft commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy:
```diff
@@ -148,8 +148,12 @@ public class AzureComputeClient extends AzureBaseClient {
     } catch (CloudException e) {
-      // treat exception as a http 404 return (resource not found)
-      log.warn("ServerGroup: ${e.message} (${serverGroupName} was not found)")
+      if (e.body.code == 404) {
```
I see an warning in IntelliJ: e.body.code is a String, 404  is an Integer...

scotmoor replied:
Yeah. I'm investigating

Commit:
Merge pull request #488 from ttomsu/google-redis-27

Fixes issue with order of GCE httpHealthCheck and instance health che…

Commit:
provider/docker: Defer loading _catalog until after initialization

ajordens committed:
Introduced a ReservedInstances caching agent (runs every 10 minutes)

- Updated ReservationReport caching agent to use the cached ReservedInstances data rather than calling AWS directly
- Adjusted ReservationReport caching agent to only run every minute

ajordens committed:
No need to @Autowire

ajordens committed:
Merge pull request #492 from ajordens/reserved-instances-caching-agent

Introduced a ReservedInstances caching agent (runs every 10 minutes)

lwander committed:
Merge pull request #491 from lwander/docker-defer-catalog

provider/docker: Defer loading _catalog until after initialization
